### PR TITLE
[#175897656] Add OTP delete after validation

### PR DESCRIPTION
**Pull request by AleDore**
#### List of Changes
- Add delete after OTP validation
- Add tests

#### Motivation and Context
Since a merchant could set `invalidate_otp` to `true` if the OTP must be invalidated, we must provide delete operations in Redis key-value pair for:
- OTP
- FiscalCode-OTP

See this [story](https://www.pivotaltracker.com/story/show/175897656) for further informations.

#### How Has This Been Tested?
Unit tests done.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by AleDore** [comment]
@balanza @BurnedMarshal: As balanza [suggests](https://github.com/pagopa/io-functions-commons/pull/135#issuecomment-799228950) I've added comments for both constants on [2955b4d](https://github.com/pagopa/io-functions-cgn-merchant/pull/2/commits/2955b4db513f47e9d4dbd14567d56d8f5177b0b2)

**Comment by BurnedMarshal** [comment]
> @balanza @BurnedMarshal: As balanza [suggests](https://github.com/pagopa/io-functions-commons/pull/135#issuecomment-799228950) I've added comments for both constants on [2955b4d](https://github.com/pagopa/io-functions-cgn-merchant/pull/2/commits/2955b4db513f47e9d4dbd14567d56d8f5177b0b2)

Thanks @AleDore. We should add the comments even on `io-functions-cgn`

File: ValidateOtp/__tests__/handler.test.ts [hunk 1]
```diff
@@ -53,6 +63,39 @@ describe("ValidateOtpHandler", () => {
...
+  it("should return an internal error if Redis delete fails", async () => {
+    deleteTaskMock.mockImplementationOnce(() =>
+      fromLeft("Cannot delete from Redis")
+    );
+    const handler = ValidateOtpHandler({} as any);
+    const response = await handler(
+      {} as any,
+      aValidationPayloadWithInvalidation
+    );
+    expect(response.kind).toBe("IResponseErrorInternal");
```
Comment (BurnedMarshal) [inline ValidateOtp/__tests__/handler.test.ts:79]:
```suggestion
    expect(deleteTaskMock).toBeCalledTimes(1);
    expect(response.kind).toBe("IResponseErrorInternal");
```

File: ValidateOtp/__tests__/handler.test.ts [hunk 2]
```diff
@@ -53,6 +63,39 @@ describe("ValidateOtpHandler", () => {
...
+  });
+
+  it("should return an internal error if OTP delete fails", async () => {
+    deleteTaskMock.mockImplementationOnce(() => taskEither.of(false));
+    const handler = ValidateOtpHandler({} as any);
+    const response = await handler(
+      {} as any,
+      aValidationPayloadWithInvalidation
+    );
+    expect(response.kind).toBe("IResponseErrorInternal");
```
Comment (BurnedMarshal) [inline ValidateOtp/__tests__/handler.test.ts:87]:
```suggestion
    expect(deleteTaskMock).toBeCalledTimes(1);
    expect(response.kind).toBe("IResponseErrorInternal");
```

File: ValidateOtp/__tests__/handler.test.ts [hunk 3]
```diff
@@ -53,6 +63,39 @@ describe("ValidateOtpHandler", () => {
...
+
+  it("should return an internal error if fiscalCode-OTP delete fails", async () => {
+    deleteTaskMock.mockImplementationOnce(() => taskEither.of(true));
+    deleteTaskMock.mockImplementationOnce(() => taskEither.of(false));
+    const handler = ValidateOtpHandler({} as any);
+    const response = await handler(
+      {} as any,
+      aValidationPayloadWithInvalidation
+    );
+    expect(response.kind).toBe("IResponseErrorInternal");
```
Comment (BurnedMarshal) [inline ValidateOtp/__tests__/handler.test.ts:96]:
```suggestion
    expect(deleteTaskMock).toBeCalledTimes(2);
    expect(response.kind).toBe("IResponseErrorInternal");
```

File: ValidateOtp/__tests__/handler.test.ts [hunk 4]
```diff
@@ -70,4 +113,18 @@ describe("ValidateOtpHandler", () => {
...
     }
   });
+
+  it("should return success if otp validation and delete succeed", async () => {
+    const handler = ValidateOtpHandler({} as any);
+    const response = await handler(
+      {} as any,
+      aValidationPayloadWithInvalidation
+    );
+    expect(response.kind).toBe("IResponseSuccessJson");
```
Comment (BurnedMarshal) [inline ValidateOtp/__tests__/handler.test.ts:131]:
```suggestion
    expect(deleteTaskMock).toBeCalledTimes(2);
    expect(deleteTaskMock).toHaveBeenNthCalledWith(
      1,
      expect.any(Object),
      `${OTP_PREFIX}${aValidationPayloadWithInvalidation.otp_code}`
    );
    expect(deleteTaskMock).toHaveBeenNthCalledWith(
      2,
      expect.any(Object),
      `${OTP_FISCAL_CODE_PREFIX}${anOtpPayload.fiscalCode}`
    );
    expect(response.kind).toBe("IResponseSuccessJson");
```

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: ValidateOtp/handler.ts [hunk 5]
```diff
@@ -87,13 +135,25 @@ export function ValidateOtpHandler(
       .mapLeft<IResponseErrorInternal | IResponseErrorNotFound>(() =>
         ResponseErrorInternal("Cannot validate OTP Code")
```
Comment (BurnedMarshal) [inline ValidateOtp/handler.ts]: On the errors scenarios, we haven't any log. Could be useful to log for debugging reason. Make sure doesn't log confidential data like text plain OTP.

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: ValidateOtp/handler.ts [hunk 6]
```diff
@@ -128,12 +129,15 @@ const invalidateOtp = (
...
+  logPrefix: string = "ValidateOtpHandler"
 ): IGetValidateOtpHandler {
-  return async (_, payload) =>
-    retrieveOtp(redisClient, payload.otp_code)
-      .mapLeft<IResponseErrorInternal | IResponseErrorNotFound>(() =>
-        ResponseErrorInternal("Cannot validate OTP Code")
+  return async (context, payload) => {
+    const otpCode = payload.otp_code;
+    const errorLogMapping = mapWithPrivacyLog(context, logPrefix, otpCode);
+    return retrieveOtp(redisClient, otpCode)
```
Comment (BurnedMarshal) [inline ValidateOtp/handler.ts]: The new constant seems useless, can we remove it?
```suggestion
    const errorLogMapping = mapWithPrivacyLog(
      context,
      logPrefix,
      payload.otp_code
    );
    return retrieveOtp(redisClient, payload.otp_code)
```

File: utils/logging.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { Context } from "@azure/functions";
+
+export const mapWithPrivacyLog = (
+  context: Context,
+  logPrefix: string,
+  stringToObfuscate?: string
```
Comment (BurnedMarshal) [inline utils/logging.ts]: Can we use a `NonEmptyString` type?
```suggestion
  stringToObfuscate?: NonEmptyString
```

**Review by BurnedMarshal — APPROVED** [review]
LGTM apart two minor changes

**Pull request merged**